[Jonathan Lam] iP

JonLamy opened the pull request:
# Duke
> Nobody wants tasks. But since we don't have a choice, why not rely on Duke for some help?

Duke is a task manager with features:
- Text-based
- _Easy_ to learn
- **Foolproof**
- and only ~~$1.99~~ **FREE**!

Download it [here](https://github.com/JonLamy/ip/releases) 🥳

Features:
- [x] Task management
- [ ] Motivation (soon)
---
If you love Java, here's a peek at Duke's `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Have fun!

Commit:
Add Gradle support

JonLamy committed:
Add simple echo.

JonLamy committed:
Move scanner.

JonLamy committed:
Task adding and check toggle.

JonLamy committed:
Added new task types.

JonLamy committed:
Basic setup for automating testing.

JonLamy committed:
Basic error handling.

JonLamy committed:
Improved error handling

JonLamy committed:
Redesign logic and renamed project.

JonLamy committed:
Refined automated testing.

JonLamy committed:
Prettify outputs.

JonLamy committed:
Follow Java coding standard

Modified class/enum names and variables to match naming conventions.
Added all project files to their relavant packages
like `task` or `exception`.
Added relavant class header comments and Javadocs for understandability.

Additionally, improved readability by abstracting setCommand function.

JonLamy committed:
Add read file function

MrRobot will search for existing task data on start.
If file is not found, it will create a blank file.
If file is found, it will load it into its memory.

JonLamy committed:
Exclude data files from tracking

JonLamy committed:
Level-7

Implement both reading and writing to hard-disk functionality.
Mr. Robot will search and load existing task data on start.

When tasklist is changed during runtime,
it will write changes to the file.

Task data in memory will be saved again on exit.

JonLamy committed:
Level-8

Mr Robot can recognise dates given for deadline tasks,
provided they are given in the right format.

If input format by user is not supported,
an error and example will be shown, and user will be prompted again.

JonLamy committed:
Merge branch 'branch-Level-8'

Resolved constructor differences in Deadline class.

JonLamy committed:
Fix error on saving deadline

JonLamy committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

JonLamy committed:
Implement OOP paradigm

Abstracted Ui, Storage, Parser, and Tasklist classes.
Commands also get their own classes.

Exceptions will print more useful responses to the user.

JonLamy committed:
Add JUnit tests

Added JUnit support, and ParserTest, StorageTest to
test for the Parser and Storage classes.

JonLamy committed:
Support JAR execution

Duke will create a local file when executed.

JonLamy committed:
A-JavaDoc

Add JavaDoc comments to code

JonLamy committed:
A-CodingStandard

Tweak code to comly with Java coding standard.

JonLamy committed:
Level-9

Allows user to find tasks with descriptions containing specified keyword.

JonLamy committed:
Add fallthrough comments

JonLamy committed:
Merge branch 'branch-A-CodingStandard'

JonLamy committed:
Merge branch 'branch-Level-9'

JonLamy committed:
Import FindCommand

JonLamy committed:
Add JavaDoc for FindCommand

JonLamy committed:
A-CheckStyle

Use Intellij to detect coding style violations.

maxng17 commented on src/main/java/ip/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     * @param s Scanner to be stored.
+     */
+    public void load(Scanner s) {
+        inputLine = s;
```
Might be better to have it as Scanner inputLine so this line written as inputLine = inputLine

maxng17 commented on src/main/java/ip/Ui.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class Ui {
+    private Scanner input = new Scanner(System.in);
+    private final String greeting = "  ___          ___  \n"
```
Following the 2103T code standards, greeting should be all caps since its a constant.

maxng17 commented on src/main/java/ip/Ui.java:
```diff
@@ -0,0 +1,59 @@
+                                   + " (o o)        (o o) \n"
+                                   + "(  V  ) DUKE (  V  )\n"
+                                   + "--m-m----------m-m--";
+    private final String bye = " ^ ^         \n"
```
All caps for bye like above.

maxng17 commented on src/main/java/ip/exception/BadDeadline.java:
```diff
@@ -0,0 +1,23 @@
+    public String toString() {
+        if (deadline.isEmpty()) {
+            return "Deadline not specified for event!\n"
+                   + "Correct example: `deadline Homework /by 2/5/2022 2359`";
```
Might be better if the example is changed to DD/MM/YYYY 2359 or 31/8/2022 since the current example now, I can interpret it as DD/MM... or MM/DD making it abit confusing

chanyuk replied:
I agree! Or maybe another way would be to directly give instructions on formatting the date to users?

maxng17 commented on src/main/java/ip/exception/BadDeadline.java:
```diff
@@ -0,0 +1,23 @@
+                   + "Correct example: `deadline Homework /by 2/5/2022 2359`";
+        } else {
+            return "Deadline specified: \"" + deadline + "\" has incorrect formatting.\n"
+                   + "Correct example: `deadline Homework /by 2/5/2022 2359`";
```
Same idea here as above

maxng17 commented on src/main/java/ip/exception/DukeException.java:
```diff
@@ -0,0 +1,11 @@
+ */
+public class DukeException extends Exception {
+
+    public DukeException() {
```
No real need for this constructor right just like the IndexNotSpecified, unless I am wrong

maxng17 reviewed (commented):
Overall pretty good. Just some nitpicking from me cause I really couldn't find much styling errors haha. 👍

JonLamy committed:
Level-10

Add GUI to Duke implemented with JavaFX.
Removed the CLI UI from Duke. Users will now interact through the GUI exclusively.

chanyuk commented on src/main/java/ip/exception/BadTimespan.java:
```diff
@@ -0,0 +1,23 @@
+                   + "Correct example: `event Meeting /at Tuesday, 2-3pm`";
+        } else {
+            return "Timespan specified: \"" + timespan + "\" has incorrect formatting.\n"
+                   + "Correct example: `event Meeting /at Tuesday, 2-3pm`";
```
Maybe you can try including a more generalised example for formatting as suggested in your BadDeadline.java file?

chanyuk reviewed (commented):
Overall it was good! The error messages was clear and your classes seem well-abstracted.

JonLamy committed:
Assert add command

Add assertion in AddCommand::execute()

To ensure that task type has been specified by the program.

JonLamy committed:
A-Assertions

AddCommand and EditCommand

Assert that type of add/edit command exists.

Get response

Duke will notify user when an empty input is provided,
instead of quitting the app.

JonLamy committed:
Improve code quality

Enhance readability by practicing SLAP
and removing instances of deep-nested code.

JonLamy committed:
Merge pull request #1 from JonLamy/branch-A-Assertions

Assertion testing command invariants

JonLamy committed:
Include BadLineFormat

JonLamy committed:
Merge branch 'master' into branch-A-CodeQuality

JonLamy committed:
Merge pull request #2 from JonLamy/branch-A-CodeQuality

Improve code quality

JonLamy committed:
C-Undo

User cannot undo changes to the tasklist (example: delete).

This risks important task information being lost accidentally.

The undo command reverts tasklist to previous state.

The undo command cannot be undid.

Furthermore, storage function to save/update storage file
has been moved to command execution,
prevents unnecessary file writing on commands such as:
list, find, bye, undo.

JonLamy committed:
Use GitHub Actions to setup CI

Added sample .yml file provided by the CS2103T module.

JonLamy committed:
Improve comments

Changed grammar to follow standard.
E.g. "Parse" -> "Parses"

Add header comments to TaskList.java

Changed parameter description for Duke constructor.
"Path of file to ..." -> "Path of directory to ..."

Add parameter descriptions to 2 methods in Storage.java

Removed debugging print statements.

JonLamy committed:
Polish GUI, add personality, and improve JUnit tests

GUI cuts off text when it exceeds specified length.

This prevents Duke from showing all tasks when the task
list is more than 3 tasks long.

Style settings for the GUI have been changed so that
the text containers will grow to fit the length of
responses.

StorageTest.java was redundant after adding error handling
to the Storage class.

Test class has been removed, and TaskListTest.java has
instead been added.

Exception message for bad formatted input is not clear.

Date format example (2/5/2022) is given, but day and month
ordering is ambiguous.

Clarified by adding an additional format tip.

JonLamy committed:
Add product screenshot

Added screenshot of Ui.

Removed GitHub actions workflow.

JonLamy committed:
Add user guide

JonLamy committed:
Fix import ordering in test files

JonLamy committed:
Display bad formatted lines to command line

When bad line format encountered in task data file,
Duke would skip and return empty list.

This results in all data lost even if one line is wrong.

Duke will now collate all bad lines and keep good lines.
Good lines will be put in the task list when loading.
Bad lines shown to terminal on startup.

JonLamy committed:
Removed redundant readme info

Features and Usage section had significant overlap.

Removed the features section,
and added and about section for more concise overview.
